### PR TITLE
Subtypes of state in typed persistence runThen Java API 

### PR DESCRIPTION
**Pull request by johanandren**
Fixes half of #25737

The Scala part I'll do separately because the suggestion in the issue didn't work out, while this is good to go IMO.

### CODE REVIEW

**Review by patriknw — APPROVED** [review]
javadsl part LGTM

**Comment by johanandren** [comment]
Failure is this: https://github.com/akka/akka/issues/26256

I'll workaround for now

**Comment by johanandren** [comment]
Meh, Scala 2.11 and lambdas
```
[error] /home/travis/build/akka/akka/akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorJavaDslTest.java:621:1: method intercept in class akka.testkit.EventFilter cannot be applied to given types;
[error]   required: scala.Function0<T>,akka.actor.ActorSystem
[error]   found: ()->{ c.te[...]ll; },akka.actor.ActorSystem
[error]   reason: cannot infer type-variable(s) T
[error]     (argument mismatch; scala.Function0 is not a functional interface
[error]       multiple non-overriding abstract methods found in interface scala.Function0)
[error]         .intercept(
```

File: akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorJavaDslTest.java [hunk 1]
```diff
@@ -559,4 +565,68 @@ public Incremented fromJournal(Wrapper wrapper) {
...
+  @Test
+  public void failOnIncorrectExpectedStateForThenRun() {
+    TestProbe<String> probe = testKit.createTestProbe();
+    ActorRef<String> c =
+        testKit.spawn(
+            new IncorrectExpectedStateForThenRun(probe.getRef(), new PersistenceId("foiesftr")));
+
+    probe.expectMessage("started!"); // workaround for #26256
+
+    new EventFilter(Logging.Error.class, Adapter.toUntyped(testKit.system()))
```
Comment (chbatey) [inline akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorJavaDslTest.java]: Maybe a more specific filter as there are some other error logs from this class:

```
[ERROR] [02/04/2019 09:55:36.435] [PersistentActorJavaDslTest-akka.actor.default-dispatcher-9] [akka://PersistentActorJavaDslTest/user] akka.persistence.RecoveryPermitter$RecoveryPermitGranted$ cannot be cast to akka.persistence.typed.javadsl.PersistentActorJavaDslTest$Command
java.lang.ClassCastException: akka.persistence.RecoveryPermitter$RecoveryPermitGranted$ cannot be cast to akka.persistence.typed.javadsl.PersistentActorJavaDslTest$Command
	at akka.persistence.typed.javadsl.PersistentActorJavaDslTest$6.aroundReceive(PersistentActorJavaDslTest.java:466)
	at akka.actor.typed.internal.InterceptorImpl.receive(InterceptorImpl.scala:73)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:421)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:394)
	at akka.actor.typed.internal.adapter.ActorAdapter.akka$actor$typed$internal$adapter$ActorAdapter$$handleMessage(ActorAdapter.scala:82)
	at akka.actor.typed.internal.adapter.ActorAdapter$$anonfun$running$1.applyOrElse(ActorAdapter.scala:78)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:39)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:592)
	at akka.actor.ActorCell.invoke(ActorCell.scala:561)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

```

**Review by chbatey — APPROVED** [review]
LGTM

File: akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorJavaDslTest.java [hunk 2]
```diff
@@ -559,4 +565,68 @@ public Incremented fromJournal(Wrapper wrapper) {
...
+      return 1;
+    }
+
+    @Override
+    public CommandHandler<String, String, Object> commandHandler() {
+      return newCommandHandlerBuilder()
+          .forAnyState()
+          .matchCommand(
+              msg -> msg.equals("expect wrong type"),
+              (context) ->
```
[the remark this answers is not in the document]
Comment (johanandren) [reply akka-persistence-typed/src/test/java/akka/persistence/typed/javadsl/PersistentActorJavaDslTest.java]: The funky formatter did it!

**Comment by patriknw** [comment]
some minor comment, then merge

**Pull request merged**